In [77]:
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname('__file__'), '..', "scripts"))
# sys.path.insert(0, "C:/Disertation/UoB-GeneTraceAI-25-26/src/scripts")
# Navigate up to 'src' then down to 'scripts'
from pathlib import Path
print(Path.cwd())
print(sys.path)

import pandas as pd
import numpy as np
from src.scripts.data_utils import load_raw_parquets, load_clean_parquets
from src.scripts.data_utils import save_dataframes_to_parquet

D:\MSc Data Science\Dissertation\AstraZeneca\UoB-GeneTraceAI-25-26
['..\\scripts', '..\\scripts', 'C:\\Users\\Thiruvel A P\\AppData\\Local\\Programs\\Python\\Python312\\python312.zip', 'C:\\Users\\Thiruvel A P\\AppData\\Local\\Programs\\Python\\Python312\\DLLs', 'C:\\Users\\Thiruvel A P\\AppData\\Local\\Programs\\Python\\Python312\\Lib', 'C:\\Users\\Thiruvel A P\\AppData\\Local\\Programs\\Python\\Python312', 'D:\\MSc Data Science\\Dissertation\\AstraZeneca\\UoB-GeneTraceAI-25-26\\GTAI', '', 'D:\\MSc Data Science\\Dissertation\\AstraZeneca\\UoB-GeneTraceAI-25-26\\GTAI\\Lib\\site-packages']


In [78]:
tables = load_raw_parquets()

Loaded hpa_rna: (24315372, 6)
Loaded depmap_expr: (1495, 53961)
Loaded geo_expr: (19914, 3268)
Loaded proteomics: (375, 12559)
Loaded fusions: (184237, 30)
Loaded mutations: (1066869, 70)
Loaded cellosaurus: (152231, 17)
Loaded depmap_profiles: (3830, 5)
Loaded sample_info: (1840, 29)
Loaded geo_info: (3267, 23)
Loaded hpa_desc: (1206, 7)
Loaded metabolomics: (928, 227)
Loaded mirna: (734, 956)
Loaded signatures: (3021, 12)


In [79]:
tables_cleaned = load_clean_parquets()

Loaded hpa_rna: (24315372, 6)
Loaded depmap_expr: (1495, 53961)
Loaded geo_expr: (19914, 3268)
Loaded proteomics: (375, 12559)
Loaded protein_map: (12558, 3)
Loaded fusions: (184237, 32)
Loaded mutations: (1066869, 70)
Loaded cellosaurus: (152231, 17)
Loaded depmap_profiles: (3830, 5)
Loaded sample_info: (1840, 29)
Loaded geo_info: (3267, 23)
Loaded hpa_desc: (1206, 7)
Loaded metabolomics: (928, 227)
Loaded mirna: (734, 956)
Loaded signatures: (3021, 12)


In [80]:
df_cellosaurus = tables["cellosaurus"]
df_cellosaurus_clean = tables_cleaned["cellosaurus"]

df_cellosaurus_clean.head()

,cellosaurus_cell_line_name,cellosaurus_accession,secondary accession number(s),synonyms,cross-references,references identifiers,web pages,comments,str profile data,diseases,species of origin,hierarchy,originate from same individual,sex of cell,age of donor at sampling,category,date (entry history)
0,#132 pc3-1-sc-e8,cvcl_b0t9,NaN,z48-5mg-70,wikidata; q108819335,patent=ep0501779a1;,NaN,group: patented cell line. || registration: in...,NaN,NaN,ncbi_taxid=10090; ! mus musculus (mouse),cvcl_d145 ! hl-1 friendly myeloma-653,NaN,NaN,NaN,hybridoma,created: 23-09-21; last updated: 30-01-24; ver...
1,#132 pl12 sc-d1,cvcl_b0t8,NaN,z48-5mg-63,wikidata; q108819336,patent=ep0501779a1;,NaN,group: patented cell line. || registration: in...,NaN,NaN,ncbi_taxid=10090; ! mus musculus (mouse),cvcl_d145 ! hl-1 friendly myeloma-653,NaN,NaN,NaN,hybridoma,created: 23-09-21; last updated: 30-01-24; ver...
2,#15310-ln,cvcl_e548,NaN,15310-ln; ter461; ter-461; ter 461; ter479; te...,dbmhc; 48439 || ecacc; 94050311 || ihw; ihw093...,NaN,http://pathology.ucla.edu/workfiles/360cx.pdf ...,part of: 12th international histocompatibility...,NaN,NaN,ncbi_taxid=9606; ! homo sapiens (human),NaN,NaN,female,age unspecified,transformed cell line,created: 22-10-12; last updated: 30-01-24; ver...
3,#16-15,cvcl_ka96,NaN,NaN,rcb; rcb4635 || wikidata; q54422067,pubmed=25400923;,NaN,monoclonal antibody isotype: igm. || monoclona...,NaN,NaN,ncbi_taxid=10090; ! mus musculus (mouse) || nc...,cvcl_4032 ! p3x63ag8.653,NaN,NaN,NaN,hybridoma,created: 22-08-17; last updated: 21-03-23; ver...
4,#40a,cvcl_iw91,NaN,NaN,wikidata; q54422071,pubmed=28159921;,NaN,characteristics: established from parent cell ...,NaN,ncit; c21619; mouse mesothelioma,ncbi_taxid=10090; ! mus musculus (mouse),cvcl_iw90 ! 40,NaN,male,1-2m,cancer cell line,created: 15-05-17; last updated: 29-06-23; ver...


In [81]:
df_cellosaurus.columns

Index(['Identifier (cell line name)', 'Accession (CVCL_xxxx)',
       'Secondary accession number(s)', 'Synonyms', 'Cross-references',
       'References identifiers', 'Web pages', 'Comments', 'STR profile data',
       'Diseases', 'Species of origin', 'Hierarchy',
       'Originate from same individual', 'Sex of cell',
       'Age of donor at sampling', 'Category', 'Date (entry history)'],
      dtype='str')

In [82]:
df_sample_info = tables["sample_info"]
df_sample_info_cleaned = tables_cleaned["sample_info"]

df_sample_info_cleaned.head()

,depmap_id,cell_line_name,stripped_cell_line_name,ccle_name,alias,cosmicid,sex,source,rrid,wtsi_master_cell_id,...,lineage_sub_subtype,lineage_molecular_subtype,default_growth_pattern,model_manipulation,model_manipulation_details,patient_id,parent_depmap_id,cellosaurus_ncit_disease,cellosaurus_ncit_id,cellosaurus_issues
0,ach-000016,slr 21,slr21,slr21_kidney,NaN,NaN,NaN,academic lab,cvcl_v607,NaN,...,NaN,NaN,NaN,NaN,NaN,pt-jnarlb,NaN,clear cell renal cell carcinoma,c4033,NaN
1,ach-000032,mhh-call-3,mhhcall3,mhhcall3_haematopoietic_and_lymphoid_tissue,NaN,NaN,female,dsmz,cvcl_0089,NaN,...,b_cell,NaN,NaN,NaN,NaN,pt-p2koyi,NaN,childhood b acute lymphoblastic leukemia,c9140,NaN
2,ach-000033,nci-h1819,ncih1819,ncih1819_lung,NaN,NaN,female,academic lab,cvcl_1497,NaN,...,nsclc_adenocarcinoma,NaN,NaN,NaN,NaN,pt-9p1wqv,NaN,lung adenocarcinoma,c3512,NaN
3,ach-000043,hs 895.t,hs895t,hs895t_fibroblast,NaN,NaN,female,atcc,cvcl_0993,NaN,...,NaN,NaN,2d: adherent,NaN,NaN,pt-rtuvzq,NaN,melanoma,c3224,NaN
4,ach-000049,hek te,hekte,hekte_kidney,NaN,NaN,NaN,academic lab,cvcl_ws59,NaN,...,NaN,NaN,NaN,immortalized,NaN,pt-qwyygr,NaN,NaN,NaN,no information is available about this cell li...


In [83]:
df_sample_info_cleaned.columns

Index(['depmap_id', 'cell_line_name', 'stripped_cell_line_name', 'ccle_name',
       'alias', 'cosmicid', 'sex', 'source', 'rrid', 'wtsi_master_cell_id',
       'sample_collection_site', 'primary_or_metastasis', 'primary_disease',
       'subtype', 'age', 'sanger_model_id', 'depmap_public_comments',
       'lineage', 'lineage_subtype', 'lineage_sub_subtype',
       'lineage_molecular_subtype', 'default_growth_pattern',
       'model_manipulation', 'model_manipulation_details', 'patient_id',
       'parent_depmap_id', 'cellosaurus_ncit_disease', 'cellosaurus_ncit_id',
       'cellosaurus_issues'],
      dtype='str')

In [84]:
df_sample_info_cleaned.shape

(1840, 29)

In [85]:
df_depmap_profiles = tables["depmap_profiles"]
df_depmap_profiles_cleaned = tables_cleaned["depmap_profiles"]


df_depmap_profiles_cleaned.head()

,profileid,modelcondition,modelid,datatype,weskit
0,pr-00utu3,mc-001131-kkjv,ach-001131,wgs,NaN
1,pr-01r7om,mc-000957-yckn,ach-000957,rna,NaN
2,pr-02xmlg,mc-002785-qo9e,ach-002785,rna,NaN
3,pr-04vvbz,mc-001289-bpdi,ach-001289,wes,ice
4,pr-09gmei,mc-000520-yim7,ach-000520,rna,NaN


In [86]:
df_depmap_profiles_cleaned.shape

(3830, 5)

In [87]:
df_depmap_profiles_cleaned.modelid.nunique()

1822

## HARMONIZATION

In [88]:
rrid_valid = df_sample_info['RRID'].str.match(r'^CVCL_', na=False)
print(f"RRID populated: {df_sample_info['RRID'].notna().sum()} / {len(df_sample_info)}")
print(f"RRID well-formed CVCL: {rrid_valid.sum()}")

RRID populated: 1818 / 1840
RRID well-formed CVCL: 1818


In [89]:
direct_match = df_sample_info['RRID'].isin(
        df_cellosaurus['Accession (CVCL_xxxx)']
    )
print(f"Direct RRID -> Accession match: {direct_match.sum()} / {len(df_sample_info)}")

Direct RRID -> Accession match: 1816 / 1840


In [90]:
needs_fallback = df_sample_info[~direct_match]
print(f"Rows needing name fallback: {len(needs_fallback)}")

Rows needing name fallback: 24


### sample_info

In [91]:
# Null value check
assert df_sample_info['DepMap_ID'].isna().sum() == 0

In [92]:
# Duplicated key check
assert df_sample_info['DepMap_ID'].duplicated().sum() == 0

### cellosaurus

In [93]:
# Null value check
assert df_cellosaurus['Accession (CVCL_xxxx)'].duplicated().sum() == 0

In [94]:
# Duplicated key check
assert df_cellosaurus['Accession (CVCL_xxxx)'].isna().sum() == 0

### depmap_profiles

In [95]:
# Null value check - ModelID
assert df_depmap_profiles["ModelID"].duplicated().sum() == 0

AssertionError: 

In [96]:
# Duplicated key check - ModelID
df_depmap_profiles["ModelID"].duplicated().sum()

np.int64(2008)

In [97]:
# Null value check - ProfileID
assert df_depmap_profiles["ProfileID"].duplicated().sum() == 0

In [98]:
# Duplicated key check - ModelID
assert df_depmap_profiles["ModelID"].isna().sum() == 0

#### ASSERTING CLEANED COLUMNS 

#### sample_info cleaned

In [99]:
df_sample_info_cleaned.columns

Index(['depmap_id', 'cell_line_name', 'stripped_cell_line_name', 'ccle_name',
       'alias', 'cosmicid', 'sex', 'source', 'rrid', 'wtsi_master_cell_id',
       'sample_collection_site', 'primary_or_metastasis', 'primary_disease',
       'subtype', 'age', 'sanger_model_id', 'depmap_public_comments',
       'lineage', 'lineage_subtype', 'lineage_sub_subtype',
       'lineage_molecular_subtype', 'default_growth_pattern',
       'model_manipulation', 'model_manipulation_details', 'patient_id',
       'parent_depmap_id', 'cellosaurus_ncit_disease', 'cellosaurus_ncit_id',
       'cellosaurus_issues'],
      dtype='str')

In [100]:
# Null value check
assert df_sample_info_cleaned['depmap_id'].isna().sum() == 0

In [101]:
# Duplicated key check
assert df_sample_info_cleaned['depmap_id'].duplicated().sum() == 0

#### cellosuarus cleaned

In [102]:
df_cellosaurus_clean.columns

Index(['cellosaurus_cell_line_name', 'cellosaurus_accession',
       'secondary accession number(s)', 'synonyms', 'cross-references',
       'references identifiers', 'web pages', 'comments', 'str profile data',
       'diseases', 'species of origin', 'hierarchy',
       'originate from same individual', 'sex of cell',
       'age of donor at sampling', 'category', 'date (entry history)'],
      dtype='str')

In [103]:
# Null value check
assert df_cellosaurus_clean['cellosaurus_accession'].duplicated().sum() == 0

In [104]:
# Duplicated key check
assert df_cellosaurus_clean['cellosaurus_accession'].isna().sum() == 0

#### depmap_profile cleaned

In [105]:
df_depmap_profiles_cleaned.columns

Index(['profileid', 'modelcondition', 'modelid', 'datatype', 'weskit'], dtype='str')

In [106]:
# Duplicated value check - ModelID
assert df_depmap_profiles_cleaned["modelid"].duplicated().sum() == 0

AssertionError: 

In [107]:
# print the sum of the Duplicated value check - ModelID
print(df_depmap_profiles_cleaned["modelid"].duplicated().sum())

2008


In [108]:
df_depmap_profiles_cleaned.shape

(3830, 5)

In [109]:
# Null key check - ModelID
assert df_depmap_profiles_cleaned["modelid"].isna().sum() == 0

In [110]:
# Null key check - ModelID
assert df_depmap_profiles_cleaned["profileid"].duplicated().sum() == 0

In [111]:
df_depmap_profiles_cleaned.loc[
    df_depmap_profiles_cleaned["modelid"].duplicated(keep=False),
]

,profileid,modelcondition,modelid,datatype,weskit
1,pr-01r7om,mc-000957-yckn,ach-000957,rna,NaN
2,pr-02xmlg,mc-002785-qo9e,ach-002785,rna,NaN
3,pr-04vvbz,mc-001289-bpdi,ach-001289,wes,ice
4,pr-09gmei,mc-000520-yim7,ach-000520,rna,NaN
7,pr-0c2lny,mc-000074-qybw,ach-000074,wes,agilent
...,...,...,...,...,...
3825,pr-zym15a,mc-000796-mvbn,ach-000796,wes,agilent
3826,pr-zytjvt,mc-000278-uc8l,ach-000278,wgs,NaN
3827,pr-zzhtvc,mc-000231-teoc,ach-000231,rna,NaN
3828,pr-zzjqoa,mc-000310-adwh,ach-000310,wes,ice


In [112]:
# Check the datatype combinations exist per modelid 
df_depmap_profiles_cleaned.groupby("modelid")["datatype"].apply(lambda x: sorted(x.unique())).value_counts()

datatype
[rna, wes]         954
[rna, wes, wgs]    365
[wes]              323
[rna, wgs]          88
[rna]               72
[wes, wgs]          12
[wgs]                8
Name: count, dtype: int64

In [113]:
# Boolean coverage flags (your existing pivot)
flags_pivoted = (
    df_depmap_profiles_cleaned
    .assign(flag=True)
    .pivot_table(
        index="modelid",
        columns="datatype",
        values="flag",
        aggfunc="any",
        fill_value=False
    )
    .reset_index()
    .rename(columns=lambda c: f"has_{c}" if c != "modelid" else c)
)

flags_pivoted.head()

datatype,modelid,has_rna,has_wes,has_wgs
0,ach-000001,True,True,True
1,ach-000002,True,True,False
2,ach-000003,True,False,False
3,ach-000004,True,True,False
4,ach-000005,True,True,True


In [114]:
# Pivot modelcondition per datatype (one column per datatype)
modelcondition_pivoted = (
    df_depmap_profiles_cleaned
    .pivot_table(
        index="modelid",
        columns="datatype",
        values="modelcondition",
        aggfunc="first"   
    )
    .reset_index()
    .rename(columns=lambda c: f"modelcondition_{c}" if c != "modelid" else c)
)

modelcondition_pivoted.head()

datatype,modelid,modelcondition_rna,modelcondition_wes,modelcondition_wgs
0,ach-000001,mc-000001-ztia,mc-000001-zu8p,mc-000001-ztia
1,ach-000002,mc-000002-8qny,mc-000002-updl,NaN
2,ach-000003,mc-000003-dwgn,NaN,NaN
3,ach-000004,mc-000004-pa3k,mc-000004-pa3k,NaN
4,ach-000005,mc-000005-aycs,mc-000005-aycs,mc-000005-aycs


In [115]:
# Pivot weskit per datatype (only meaningful for 'wes', will be NaN elsewhere)
weskit_pivoted = (
    df_depmap_profiles_cleaned
    .pivot_table(
        index="modelid",
        columns="datatype",
        values="weskit",
        aggfunc="first"
    )
    .reset_index()
    .rename(columns=lambda c: f"weskit_{c}" if c != "modelid" else c)
)

weskit_pivoted.head()

datatype,modelid,weskit_wes
0,ach-000001,agilent
1,ach-000002,agilent
2,ach-000004,ice
3,ach-000005,ice
4,ach-000006,agilent


In [116]:
df_depmap_profiles_cleaned_pivoted = (
    flags_pivoted
    .merge(modelcondition_pivoted, on="modelid", how="left")
    .merge(weskit_pivoted, on="modelid", how="left")
)

assert df_depmap_profiles_cleaned_pivoted["modelid"].duplicated().sum() == 0
print(f"Shape: {df_depmap_profiles_cleaned_pivoted.shape}")

df_depmap_profiles_cleaned_pivoted.head()

Shape: (1822, 8)


datatype,modelid,has_rna,has_wes,has_wgs,modelcondition_rna,modelcondition_wes,modelcondition_wgs,weskit_wes
0,ach-000001,True,True,True,mc-000001-ztia,mc-000001-zu8p,mc-000001-ztia,agilent
1,ach-000002,True,True,False,mc-000002-8qny,mc-000002-updl,NaN,agilent
2,ach-000003,True,False,False,mc-000003-dwgn,NaN,NaN,NaN
3,ach-000004,True,True,False,mc-000004-pa3k,mc-000004-pa3k,NaN,ice
4,ach-000005,True,True,True,mc-000005-aycs,mc-000005-aycs,mc-000005-aycs,ice


In [117]:
print(f"modelid+datatype duplicate pairs: {df_depmap_profiles_cleaned.duplicated(subset=["modelid", "datatype"]).sum()}")  

modelid+datatype duplicate pairs: 224


In [118]:
df_depmap_profiles_cleaned.loc[
    df_depmap_profiles_cleaned.duplicated(subset=["modelid", "datatype"])
]

,profileid,modelcondition,modelid,datatype,weskit
248,pr-3x4ri2,mc-001453-lwrm,ach-001453,wes,ice
504,pr-81ikhg,mc-000745-kwjq,ach-000745,wes,agilent
521,pr-8br5s3,mc-000329-gbth,ach-000329,wes,agilent
583,pr-97jle2,mc-001360-kfdj,ach-001360,wes,ice
594,pr-9bhyji,mc-000095-ucyl,ach-000095,rna,NaN
...,...,...,...,...,...
3797,pr-zr0jbj,mc-000055-ndma,ach-000055,wes,agilent
3812,pr-zv0mdh,mc-001321-4zqp,ach-001321,wes,agilent
3813,pr-zvd6kc,mc-000337-vmhg,ach-000337,rna,NaN
3818,pr-zwqrih,mc-002029-jzk3,ach-002029,wes,ice


### Key to aggregate the tables:
- depmap_profile: modelid
- cellosaurus: cellosaurus_accession
- sample_info: depmap_id, rrid

In [119]:
df_cellosaurus_clean["cellosaurus_accession"]

0         cvcl_b0t9
1         cvcl_b0t8
2         cvcl_e548
3         cvcl_ka96
4         cvcl_iw91
            ...    
152226    cvcl_zb29
152227    cvcl_zb30
152228    cvcl_a3zf
152229    cvcl_c6u7
152230    cvcl_3449
Name: cellosaurus_accession, Length: 152231, dtype: str

In [120]:
df_sample_info_cleaned["rrid"]

0       cvcl_v607
1       cvcl_0089
2       cvcl_1497
3       cvcl_0993
4       cvcl_ws59
          ...    
1835    cvcl_1810
1836    cvcl_0271
1837    cvcl_1296
1838    cvcl_5g89
1839    cvcl_s641
Name: rrid, Length: 1840, dtype: str

In [121]:
rrid_valid_id = set(df_sample_info_cleaned["rrid"]) 

print("Mapped cellosaurus_accession & rrid map ratio:", df_cellosaurus_clean[df_cellosaurus_clean["cellosaurus_accession"].isin(rrid_valid_id)]["cellosaurus_accession"].count())
print("Mis-Mapped cellosaurus_accession & rrid map ratio:", df_cellosaurus_clean[~df_cellosaurus_clean["cellosaurus_accession"].isin(rrid_valid_id)]["cellosaurus_accession"].count())

Mapped cellosaurus_accession & rrid map ratio: 1812
Mis-Mapped cellosaurus_accession & rrid map ratio: 150419


In [122]:
# --- STEP 1: Normalize both keys ---
cvcl_norm = df_cellosaurus_clean["cellosaurus_accession"].str.strip().str.upper()
rrid_norm  = df_sample_info_cleaned["rrid"].str.strip().str.upper()

cvcl_set = set(cvcl_norm)
rrid_set  = set(rrid_norm)

print("=== After normalization ===")
print("Matched :", len(cvcl_set & rrid_set))
print("rrid NOT in cellosaurus:", len(rrid_set - cvcl_set))

# --- STEP 2: Inspect the 28 unmatched rrid values ---
unmatched = rrid_set - cvcl_set
print("\nUnmatched rrid sample:", list(unmatched)[:15])

# --- STEP 3: Prefix audit ---
print("\nrrid prefix distribution:")
print(df_sample_info_cleaned["rrid"].str[:5].value_counts().head(10))

print("\ncellosaurus_accession prefix distribution:")
print(df_cellosaurus_clean["cellosaurus_accession"].str[:5].value_counts().head(10))

# --- STEP 4: Length distribution (detects padding/extra chars) ---
print("\nrrid length stats:")
print(df_sample_info_cleaned["rrid"].str.len().describe())

print("\ncellosaurus_accession length stats:")
print(df_cellosaurus_clean["cellosaurus_accession"].str.len().describe())

# --- STEP 5: Check for 'RRID:' prefix in rrid column ---
has_rrid_prefix = df_sample_info_cleaned["rrid"].str.startswith("RRID:").sum()
print(f"\nEntries with 'RRID:' prefix in sample_info: {has_rrid_prefix}")

=== After normalization ===
Matched : 1812
rrid NOT in cellosaurus: 3

Unmatched rrid sample: ['CVCL_X507', nan, 'CVCL_V618']

rrid prefix distribution:
rrid
cvcl_    1818
Name: count, dtype: int64

cellosaurus_accession prefix distribution:
cellosaurus_accession
cvcl_    152231
Name: count, dtype: int64

rrid length stats:
count    1818.0
mean        9.0
std         0.0
min         9.0
25%         9.0
50%         9.0
75%         9.0
max         9.0
Name: rrid, dtype: float64

cellosaurus_accession length stats:
count    152231.0
mean          9.0
std           0.0
min           9.0
25%           9.0
50%           9.0
75%           9.0
max           9.0
Name: cellosaurus_accession, dtype: float64

Entries with 'RRID:' prefix in sample_info: 0


In [123]:
df_sample_info_cleaned["depmap_id"]

0       ach-000016
1       ach-000032
2       ach-000033
3       ach-000043
4       ach-000049
           ...    
1835    ach-002393
1836    ach-002394
1837    ach-002395
1838    ach-002396
1839    ach-002397
Name: depmap_id, Length: 1840, dtype: str

In [124]:
df_depmap_profiles_cleaned_pivoted["modelid"]

0       ach-000001
1       ach-000002
2       ach-000003
3       ach-000004
4       ach-000005
           ...    
1817    ach-003157
1818    ach-003158
1819    ach-003159
1820    ach-003160
1821    ach-003161
Name: modelid, Length: 1822, dtype: str

In [125]:
depmap_id_valid = set(df_sample_info_cleaned["depmap_id"]) 

print("Mapped depmap_id & modelid map ratio:", df_depmap_profiles_cleaned_pivoted[df_depmap_profiles_cleaned_pivoted["modelid"].isin(depmap_id_valid)]["modelid"].count())
print("Mis-Mapped depmap_id & modelid map ratio:", df_depmap_profiles_cleaned_pivoted[~df_depmap_profiles_cleaned_pivoted["modelid"].isin(depmap_id_valid)]["modelid"].count())

Mapped depmap_id & modelid map ratio: 1749
Mis-Mapped depmap_id & modelid map ratio: 73


In [126]:
depmap_set  = set(df_sample_info_cleaned["depmap_id"].dropna().str.strip().str.lower())
modelid_set = set(df_depmap_profiles_cleaned_pivoted["modelid"].dropna().str.strip().str.lower())

# Both directions
in_sample_not_profile = depmap_set - modelid_set
in_profile_not_sample = modelid_set - depmap_set

print("sample_info NaN depmap_id     :", df_sample_info_cleaned["depmap_id"].isna().sum())
print("depmap_profiles NaN modelid   :", df_depmap_profiles_cleaned_pivoted["modelid"].isna().sum())
print("depmap_id NOT in modelid      :", len(in_sample_not_profile))
print("modelid NOT in depmap_id      :", len(in_profile_not_sample))

# Format audit
print("\nSample depmap_id prefix   :", df_sample_info_cleaned["depmap_id"].str[:4].value_counts())
print("Sample modelid prefix      :", df_depmap_profiles_cleaned_pivoted["modelid"].str[:4].value_counts())
print("\nSample depmap_id length   :", df_sample_info_cleaned["depmap_id"].str.len().describe())
print("Sample modelid length      :", df_depmap_profiles_cleaned_pivoted["modelid"].str.len().describe())

# Inspect mismatches
print("\nSample unmatched depmap_id :", list(in_sample_not_profile)[:10])
print("Sample unmatched modelid   :", list(in_profile_not_sample)[:10])

sample_info NaN depmap_id     : 0
depmap_profiles NaN modelid   : 0
depmap_id NOT in modelid      : 91
modelid NOT in depmap_id      : 73

Sample depmap_id prefix   : depmap_id
ach-    1840
Name: count, dtype: int64
Sample modelid prefix      : modelid
ach-    1822
Name: count, dtype: int64

Sample depmap_id length   : count    1840.0
mean       10.0
std         0.0
min        10.0
25%        10.0
50%        10.0
75%        10.0
max        10.0
Name: depmap_id, dtype: float64
Sample modelid length      : count    1822.0
mean       10.0
std         0.0
min        10.0
25%        10.0
50%        10.0
75%        10.0
max        10.0
Name: modelid, dtype: float64

Sample unmatched depmap_id : ['ach-001122', 'ach-002172', 'ach-002334', 'ach-001214', 'ach-002260', 'ach-001091', 'ach-002328', 'ach-002393', 'ach-001357', 'ach-001146']
Sample unmatched modelid   : ['ach-003135', 'ach-001705', 'ach-003152', 'ach-001679', 'ach-003159', 'ach-002535', 'ach-002533', 'ach-001847', 'ach-002650', 'ach-

### OVERALL DATA LOSS ACROSS THE THREE DATASET 
1. **Three datasets in scope** → `cellosaurus_clean` (152,231 rows), `sample_info_cleaned` (1,840 rows), `depmap_profiles_pivoted` (1,822 rows) all keyed on uniform, fixed-length identifiers (`cvcl_` / `ach-`).

2. **Cellosaurus ↔ sample_info** (via `cellosaurus_accession` ↔ `rrid`) → **24 records lost** 22 due to null `rrid` at source, 2 due to secondary/deprecated accessions absent from the provided dump.

3. **sample_info ↔ depmap_profiles** (via `depmap_id` ↔ `modelid`) → **91 records lost** from sample_info side, **73 records lost** from depmap_profiles side caused purely by DepMap release version mismatch between datasets.

4. **Zero format-related losses** across all three joins no case mismatches, no prefix pollution, no null keys in DepMap join, no length inconsistencies detected.

5. **Cumulative loss from sample_info anchor** → `24 (Cellosaurus) + 91 (DepMap) = 115 records` affected across both integrations (overlap unknown without three-way join).

6. **Final harmonized coverage** → Cellosaurus join retains **98.48%** (1,812/1,840); DepMap join retains **95.05%** (1,749/1,840); three-way intersection is the effective working dataset.

7. **Total irrecoverable loss** → **115 records maximum (6.25% of sample_info)** split between source nulls, snapshot version gaps, and deprecated registry entries; none resolvable within the provided data environment.

### REDUCE THE DATA LOSS

In [127]:
# Check if sample_info has name columns
print("Sample_info :", df_sample_info_cleaned.columns.tolist())
print("\n---------\n")
print("cellosaurus :",df_cellosaurus_clean.columns.tolist())

Sample_info : ['depmap_id', 'cell_line_name', 'stripped_cell_line_name', 'ccle_name', 'alias', 'cosmicid', 'sex', 'source', 'rrid', 'wtsi_master_cell_id', 'sample_collection_site', 'primary_or_metastasis', 'primary_disease', 'subtype', 'age', 'sanger_model_id', 'depmap_public_comments', 'lineage', 'lineage_subtype', 'lineage_sub_subtype', 'lineage_molecular_subtype', 'default_growth_pattern', 'model_manipulation', 'model_manipulation_details', 'patient_id', 'parent_depmap_id', 'cellosaurus_ncit_disease', 'cellosaurus_ncit_id', 'cellosaurus_issues']

---------

cellosaurus : ['cellosaurus_cell_line_name', 'cellosaurus_accession', 'secondary accession number(s)', 'synonyms', 'cross-references', 'references identifiers', 'web pages', 'comments', 'str profile data', 'diseases', 'species of origin', 'hierarchy', 'originate from same individual', 'sex of cell', 'age of donor at sampling', 'category', 'date (entry history)']


In [128]:
# Chceking the presese of the missing keys (CVCL_X507 & CVCL_V618) 
# Explode secondary accessions and match against rrid
secondary = df_cellosaurus_clean["secondary accession number(s)"].str.split("|").explode()
print(secondary[secondary.isin(["cvcl_x507", "cvcl_v618"])])

125217    cvcl_x507
Name: secondary accession number(s), dtype: str


In [129]:
print(df_cellosaurus_clean["cross-references"].head(5))

0                                 wikidata; q108819335
1                                 wikidata; q108819336
2    dbmhc; 48439 || ecacc; 94050311 || ihw; ihw093...
3                  rcb; rcb4635 || wikidata; q54422067
4                                  wikidata; q54422071
Name: cross-references, dtype: str


In [130]:
# Check 1 — secondary accession coverage
print(df_cellosaurus_clean["secondary accession number(s)"].notna().sum())
print(df_cellosaurus_clean["secondary accession number(s)"].head(3))

# Check 2 — cross-references format  
print(df_cellosaurus_clean["cross-references"].head(3))

# Check 3 — cellosaurus_ncit_id already in sample_info
print(df_sample_info_cleaned["cellosaurus_ncit_id"].notna().sum())

542
0    NaN
1    NaN
2    NaN
Name: secondary accession number(s), dtype: str
0                                 wikidata; q108819335
1                                 wikidata; q108819336
2    dbmhc; 48439 || ecacc; 94050311 || ihw; ihw093...
Name: cross-references, dtype: str
1747


In [131]:
# Explode secondary accessions → join to unmatched rrid
sec_acc = df_cellosaurus_clean[["cellosaurus_accession", "secondary accession number(s)"]].copy()
sec_acc = sec_acc.dropna(subset=["secondary accession number(s)"])
sec_acc["secondary_exploded"] = sec_acc["secondary accession number(s)"].str.split("||")
sec_acc = sec_acc.explode("secondary_exploded")
sec_acc["secondary_exploded"] = sec_acc["secondary_exploded"].str.strip()

# Debug raw value first
print(df_cellosaurus_clean.loc[125217, "secondary accession number(s)"])

# Likely delimiter is " || " not "||" — fix:
sec_acc["secondary_exploded"] = (
    sec_acc["secondary accession number(s)"]
    .str.split(r"\s*\|\|\s*", regex=True)   # handles any spacing
)
sec_acc = sec_acc.explode("secondary_exploded")
sec_acc["secondary_exploded"] = sec_acc["secondary_exploded"].str.strip().str.lower()


# Recover
recovered = df_sample_info_cleaned[
    df_sample_info_cleaned["rrid"].isin(sec_acc["secondary_exploded"])
].merge(sec_acc, left_on="rrid", right_on="secondary_exploded")

print("Recovered via secondary:", len(recovered))

cvcl_x507
Recovered via secondary: 11


In [132]:
# Cross-references format: "source; id || source; id"
def parse_cross_refs(ref_str):
    if pd.isna(ref_str): return {}
    pairs = {}
    for entry in ref_str.split("||"):
        if ";" in entry:
            source, id_ = entry.split(";", 1)
            pairs[source.strip().lower()] = id_.strip()
    return pairs

df_cellosaurus_clean["refs_parsed"] = df_cellosaurus_clean["cross-references"].apply(parse_cross_refs)

# Check all available source keys (look for depmap, cosmic, ccle)
all_sources = set()
df_cellosaurus_clean["refs_parsed"].apply(lambda d: all_sources.update(d.keys()))
print(sorted(all_sources)) 

# Extract depmap ACH IDs from cellosaurus cross-references
def extract_depmap(ref_str):
    if pd.isna(ref_str): return None
    for entry in ref_str.split("||"):
        if "depmap" in entry.lower() and ";" in entry:
            return entry.split(";", 1)[1].strip().lower()
    return None

df_cellosaurus_clean["depmap_xref"] = df_cellosaurus_clean["cross-references"].apply(extract_depmap)

print("Cellosaurus rows with depmap xref:", df_cellosaurus_clean["depmap_xref"].notna().sum())

# Direct bridge: cellosaurus → depmap_profiles (bypass sample_info gap)
bridge = df_cellosaurus_clean[["cellosaurus_accession", "depmap_xref"]].dropna()
recovered_depmap = bridge[bridge["depmap_xref"].isin(
    df_depmap_profiles_cleaned_pivoted["modelid"].str.lower()
)]
print("Cellosaurus → depmap_profiles direct match:", len(recovered_depmap))

['4dn', 'abcam', 'abcd', 'abm', 'addexbio', 'arrayexpress', 'atcc', 'bcgo', 'bcrc', 'bcrj', 'bei_resources', 'biogrid_orcs_cell_line', 'biosample', 'biosamples', 'bto', 'cancercelllines', 'cancertools', 'cba', 'cclv', 'ccrid', 'cctcc', 'cell_biolabs', 'cell_model_passport', 'cgh-db', 'chembl-cells', 'chembl-targets', 'cldb', 'clo', 'cls', 'colonatlas', 'coriell', 'cosmic', 'cosmic-clp', 'dbgap', 'dbmhc', 'depmap', 'dgrc', 'discoverx', 'dshb', 'dsmz', 'dsmzcelldive', 'ebisc', 'ecacc', 'efo', 'ega', 'encode', 'estdab', 'fcdi', 'fcs-free', 'flybase_cell_line', 'gdsc', 'genecopoeia', 'geo', 'hipsci', 'hivreagentprogram', 'horizon_discovery', 'hpscreg', 'iarc_tp53', 'ibrc', 'iclc', 'icldb', 'igrhcellid', 'igsr', 'ihw', 'imanis', 'innoprot', 'ipd-imgt/hla', 'iscr', 'izsler', 'jcrb', 'kcb', 'kclb', 'kerafast', 'kyinno', 'ligea', 'limore', 'lincs_hms', 'lincs_ldp', 'lonza', 'mccl', 'mesh', 'metabolights', 'millipore', 'mmrrc', 'ncbi_iran', 'nci-dtp', 'nhcdr', 'nihhesc', 'nises', 'nrfc', 'perki

In [133]:
# Normalize both name columns
sample_names = df_sample_info_cleaned[df_sample_info_cleaned["rrid"].isna()].copy()
sample_names["name_norm"] = sample_names["cell_line_name"].str.strip().str.lower()

cvcl_names = df_cellosaurus_clean.copy()
cvcl_names["name_norm"] = cvcl_names["cellosaurus_cell_line_name"].str.strip().str.lower()

# Tier 2: exact name match
t2 = sample_names.merge(cvcl_names, on="name_norm")
print("Recovered via name match:", len(t2))

# Your concern is right — names are NOT guaranteed unique
name_counts = df_cellosaurus_clean["cellosaurus_cell_line_name"].str.strip().str.lower().value_counts()
duplicate_names = name_counts[name_counts > 1]
print("Duplicate names in cellosaurus:", len(duplicate_names))

# Only accept name match if it maps to exactly ONE cvcl
t2_safe = t2[~t2["name_norm"].isin(duplicate_names.index)]
print("Safe name recoveries:", len(t2_safe))

Recovered via name match: 8
Duplicate names in cellosaurus: 69
Safe name recoveries: 8


In [134]:
# 1. Confirm 3-tier cellosaurus join gives 1,831
# 2. Use cellosaurus depmap_xref as direct bridge
df_final = df_cellosaurus_clean[["cellosaurus_accession", "depmap_xref"]].merge(
    df_depmap_profiles_cleaned_pivoted,
    left_on="depmap_xref", right_on="modelid",
    how="outer"
)
print(df_final.shape)

# 3. Check remaining 9 unresolved in cellosaurus side
print(df_sample_info_cleaned[
    ~df_sample_info_cleaned["depmap_id"].isin(df_final["depmap_xref"])
]["cell_line_name"].tolist())

(152285, 10)
['a673stag2ko16', 'a673stag2ko45', 'a673stag2nt14', 'a673stag2nt23', 'myla', 'u-251 mg', 'hcc827gr', 'ss1a', 'gbm001', 'sc-1', 'a375_rpmi', 'rvh421_rpmi', 'nci-h292', nan, 'pl18', 'rh-1', 'ctv-1', nan]


In [135]:
# Parse cosmic & cell_model_passport from cellosaurus cross-references
def extract_xref(ref_str, source):
    if pd.isna(ref_str): return None
    for entry in ref_str.split("||"):
        if source in entry.lower() and ";" in entry:
            return entry.split(";", 1)[1].strip()
    return None

df_cellosaurus_clean["cosmic_xref"]   = df_cellosaurus_clean["cross-references"].apply(lambda x: extract_xref(x, "cosmic"))
df_cellosaurus_clean["sanger_xref"]   = df_cellosaurus_clean["cross-references"].apply(lambda x: extract_xref(x, "cell_model_passport"))

# Match against sample_info
cosmic_match  = df_sample_info_cleaned["cosmicid"].astype(str).isin(df_cellosaurus_clean["cosmic_xref"])
sanger_match  = df_sample_info_cleaned["sanger_model_id"].isin(df_cellosaurus_clean["sanger_xref"])

print("Recoverable via COSMIC :", cosmic_match.sum())
print("Recoverable via Sanger :", sanger_match.sum())

Recoverable via COSMIC : 859
Recoverable via Sanger : 1838


#### INSIGHTS FROM THE DATA RECOVERY METHODS

Table 1: Cellosaurus ↔ Sample_info
(via cellosaurus_accession ↔ rrid → sanger_model_id)

Dataset sizes → Cellosaurus: 152,231 rows, Sample_info: 1,840 rows (anchor)
Initial RRID exact match → 1,812 / 1,840 = 98.48% via rrid ↔ cellosaurus_accession
RRID null mapping → 22 / 1,840 = 1.20% have null rrid at source no accession recorded at data collection; unresolvable via RRID chain
Deprecated/secondary accession → 2 valid-format RRIDs (CVCL_X507, CVCL_V618) absent from primary dump; secondary accession explode recovered +1 (CVCL_X507), 1 permanently lost
Name-based fallback → safe exact name match (excluding 69 duplicate names in Cellosaurus) recovered +8 / 22 null-rrid rows
Sanger ID mapping → sanger_model_id ↔ cell_model_passport xref recovered 1,838 / 1,840 = 99.89% supersedes RRID as primary key
Hard floor loss → 2 / 1,840 = 0.11% both are NaN rows with no identifier; irrecoverable

Resolution summary → initial loss was 28 / 1,840 = 1.52% (28 unmatched); multi-tier recovery (secondary accession +1, name fallback +8, Sanger primary key +17) resolved 26 / 28 = 92.86% of initial loss → final matching rate improved from 98.48% → 99.89% (1,838 / 1,840) with only 2 irrecoverable NaN rows (0.11%) remaining.

Table 2: Sample_info ↔ DepMap Profiles
(via depmap_id ↔ modelid)

Dataset sizes → Sample_info: 1,840 rows, DepMap Profiles: 1,822 rows
Initial depmap_id exact match → 1,749 / 1,840 = 95.05% via depmap_id ↔ modelid
Directional loss → 91 depmap_ids in sample_info absent from profiles (4.95%); 73 modelids in profiles absent from sample_info — bidirectional gap
Root cause → pure DepMap release version mismatch — zero format issues (identical ach- prefix, length=10, zero nulls on both sides)
Cellosaurus xref bridge → depmap field in Cellosaurus cross-references yielded 1,893 entries; direct Cellosaurus → DepMap match = 1,768 / 1,822 = 97.04% — recovers +19 over initial join
Residual loss → 72 / 1,840 = 3.91% — cell lines present in sample_info but absent from both depmap_profiles and cellosaurus xref; version gap irrecoverable without unified DepMap release

Resolution summary → initial loss was 91 / 1,840 = 4.95% (91 unmatched from sample_info side); Cellosaurus depmap xref bridge recovered +19 records = 20.88% of initial loss resolved → final matching rate improved from 95.05% → 96.09% (1,768 / 1,840) with 72 irrecoverable records (3.91%) remaining due to release version gap.

### Load the refernces

In [136]:
gene_lookup      = pd.read_parquet(os.path.join("reference/", "gene_lookup.parquet"))
cell_line_lookup = pd.read_parquet(os.path.join("reference/", "cell_line_lookup.parquet"))

In [137]:
gene_lookup.head()

,ensg_id,hgnc_symbol,prev_symbols,alias_symbols,hgnc_id,biotype
0,ENSG00000121410,A1BG,NaN,NaN,HGNC:5,protein_coding
1,ENSG00000148584,A1CF,NaN,ACF|ASP|ACF64|ACF65|APOBEC1CF,HGNC:24086,protein_coding
2,ENSG00000175899,A2M,NaN,FWP007|S863-7|CPAMD5,HGNC:7,protein_coding
3,ENSG00000166535,A2ML1,CPAMD9,FLJ25179|p170,HGNC:23336,protein_coding
4,ENSG00000184389,A3GALT2,A3GALT2P,IGBS3S|IGB3S,HGNC:30005,protein_coding


In [138]:
cell_line_lookup.head()

,model_id,cell_line_name,stripped_cell_line_name,cvcl_accession,synonyms,rrid
0,ACH-000016,SLR 21,SLR21,CVCL_V607,SLR 21,CVCL_V607
1,ACH-000032,MHH-CALL-3,MHHCALL3,CVCL_0089,Mhh-Call 3|MHH cALL 3|MHH-CALL3|MHH-cALL3|MHHC...,CVCL_0089
2,ACH-000033,NCI-H1819,NCIH1819,CVCL_1497,H1819|H-1819|NCIH1819,CVCL_1497
3,ACH-000043,Hs 895.T,HS895T,CVCL_0993,Hs-895-T|Hs 895 T|Hs895.T|Hs895T|HS895T,CVCL_0993
4,ACH-000049,HEK TE,HEKTE,CVCL_WS59,HEK-TE|HEKTE,CVCL_WS59


#### TRACK A HARMONIZATION WITH THE LOOKUP TABLE

### Key to aggregate the tables:
- depmap_profile: modelid
- cellosaurus: cellosaurus_accession
- sample_info: depmap_id, rrid

In [139]:
df_cellosaurus_clean.columns

Index(['cellosaurus_cell_line_name', 'cellosaurus_accession',
       'secondary accession number(s)', 'synonyms', 'cross-references',
       'references identifiers', 'web pages', 'comments', 'str profile data',
       'diseases', 'species of origin', 'hierarchy',
       'originate from same individual', 'sex of cell',
       'age of donor at sampling', 'category', 'date (entry history)',
       'refs_parsed', 'depmap_xref', 'cosmic_xref', 'sanger_xref'],
      dtype='str')

In [140]:
df_depmap_profiles_cleaned_pivoted.columns

Index(['modelid', 'has_rna', 'has_wes', 'has_wgs', 'modelcondition_rna',
       'modelcondition_wes', 'modelcondition_wgs', 'weskit_wes'],
      dtype='str', name='datatype')

In [141]:
df_sample_info_cleaned.columns

Index(['depmap_id', 'cell_line_name', 'stripped_cell_line_name', 'ccle_name',
       'alias', 'cosmicid', 'sex', 'source', 'rrid', 'wtsi_master_cell_id',
       'sample_collection_site', 'primary_or_metastasis', 'primary_disease',
       'subtype', 'age', 'sanger_model_id', 'depmap_public_comments',
       'lineage', 'lineage_subtype', 'lineage_sub_subtype',
       'lineage_molecular_subtype', 'default_growth_pattern',
       'model_manipulation', 'model_manipulation_details', 'patient_id',
       'parent_depmap_id', 'cellosaurus_ncit_disease', 'cellosaurus_ncit_id',
       'cellosaurus_issues'],
      dtype='str')

In [142]:
cell_line_lookup.columns

Index(['model_id', 'cell_line_name', 'stripped_cell_line_name',
       'cvcl_accession', 'synonyms', 'rrid'],
      dtype='str')

### Track A: Cell Line Lookup Integration

- Architecture: hub-and-spoke, NOT pairwise outer joins.
    - Anchor  = sample_info            (owns canonical model_id, PK never null/dup)
    - Spoke 1 = cellosaurus            (resolved via 5-step cascade, NOT direct join)
    - Spoke 2 = depmap_profiles        (already ACH-keyed, direct left join bridge table for)

- Set theory per step:
    resolved_i   = unmatched_i * candidate_source_i
    unmatched_{i+1} = unmatched_i \\ resolved_i          (strictly shrinking, first-match-wins)

- Inputs expected in scope:
    df_sample_info_cleaned, df_cellosaurus_clean, df_depmap_profiles_cleaned_pivoted


In [143]:
# sample_info.depmap_id IS the canonical model_id.
# No cascade needed here — this table owns the PK.
cell_line_lookup = (
    df_sample_info_cleaned[
        ["depmap_id", "cell_line_name", "stripped_cell_line_name", "rrid"]
    ]
    .rename(columns={"depmap_id": "model_id"})
    .dropna(subset=["model_id"])          # PK never null — §2
    .drop_duplicates(subset=["model_id"]) # PK never duplicated — §2
    .reset_index(drop=True)
)

cell_line_lookup["cvcl_accession"] = pd.NA
cell_line_lookup["synonyms"] = pd.NA
cell_line_lookup["_resolution_step"] = pd.NA   # audit column, drop before final export

In [144]:

# unmatched_0 = all of cell_line_lookup (cvcl_accession is NA for every row)
ces = df_cellosaurus_clean.copy()
def _normalize_cvcl(series: pd.Series) -> pd.Series:
    """Extract CVCL_XXXXXXX pattern from RRID/accession strings for robust matching."""
    return series.astype(str).str.extract(r"(CVCL_\d+)", expand=False)

def _strip_name(series: pd.Series) -> pd.Series:
    return series.astype(str).str.upper().str.replace(r"[-_\s]", "", regex=True)

cell_line_lookup["_rrid_cvcl"] = _normalize_cvcl(cell_line_lookup["rrid"])

In [145]:
# ---- Step 2a: direct RRID match --------------------------------------
unmatched = cell_line_lookup["cvcl_accession"].isna()
lookup = ces.dropna(subset=["cellosaurus_accession"]).drop_duplicates("cellosaurus_accession")
print(lookup.columns)
merged = cell_line_lookup.loc[unmatched, "_rrid_cvcl"].map(
    lookup.set_index("cellosaurus_accession", drop=False)[["cellosaurus_accession", "synonyms"]].to_dict("index")
)
resolved_mask = unmatched & merged.notna()
cell_line_lookup.loc[resolved_mask, "cvcl_accession"] = merged[merged.notna()].apply(lambda d: d["cellosaurus_accession"])
cell_line_lookup.loc[resolved_mask, "synonyms"] = merged[merged.notna()].apply(lambda d: d["synonyms"])
cell_line_lookup.loc[resolved_mask, "_resolution_step"] = "direct_rrid"

# ---- Step 2b: secondary accession match -------------------------------
unmatched = cell_line_lookup["cvcl_accession"].isna()
sec = ces[["cellosaurus_accession", "secondary accession number(s)", "synonyms"]].copy()
sec["secondary accession number(s)"] = sec["secondary accession number(s)"].astype(str).str.split("|")
sec = sec.explode("secondary accession number(s)")
sec["secondary accession number(s)"] = sec["secondary accession number(s)"].str.strip()
sec = sec.dropna(subset=["secondary accession number(s)"]).drop_duplicates("secondary accession number(s)")  # first match wins
sec_map = sec.set_index("secondary accession number(s)")[["cellosaurus_accession", "synonyms"]].to_dict("index")

merged = cell_line_lookup.loc[unmatched, "_rrid_cvcl"].map(sec_map)
resolved_mask = unmatched & merged.notna()
cell_line_lookup.loc[resolved_mask, "cvcl_accession"] = merged[merged.notna()].apply(lambda d: d["cellosaurus_accession"])
cell_line_lookup.loc[resolved_mask, "synonyms"] = merged[merged.notna()].apply(lambda d: d["synonyms"])
cell_line_lookup.loc[resolved_mask, "_resolution_step"] = "secondary_accession"

# ---- Step 2c: name match (exact, then stripped) ------------------------
unmatched = cell_line_lookup["cvcl_accession"].isna()
name_lookup = ces.dropna(subset=["cellosaurus_cell_line_name"]).drop_duplicates("cellosaurus_cell_line_name")
name_map = name_lookup.set_index("cellosaurus_cell_line_name")[["cellosaurus_accession", "synonyms"]].to_dict("index")

merged = cell_line_lookup.loc[unmatched, "cell_line_name"].map(name_map)
resolved_mask = unmatched & merged.notna()
cell_line_lookup.loc[resolved_mask, "cvcl_accession"] = merged[merged.notna()].apply(lambda d: d["cellosaurus_accession"])
cell_line_lookup.loc[resolved_mask, "synonyms"] = merged[merged.notna()].apply(lambda d: d["synonyms"])
cell_line_lookup.loc[resolved_mask, "_resolution_step"] = "name_match"

unmatched = cell_line_lookup["cvcl_accession"].isna()
name_lookup["_stripped_name"] = _strip_name(name_lookup["cellosaurus_cell_line_name"])
name_lookup_stripped = name_lookup.drop_duplicates("_stripped_name")
stripped_map = name_lookup_stripped.set_index("_stripped_name")[["cellosaurus_accession", "synonyms"]].to_dict("index")

merged = _strip_name(cell_line_lookup.loc[unmatched, "stripped_cell_line_name"]).map(stripped_map)
resolved_mask = unmatched & merged.notna()
cell_line_lookup.loc[resolved_mask, "cvcl_accession"] = merged[merged.notna()].apply(lambda d: d["cellosaurus_accession"])
cell_line_lookup.loc[resolved_mask, "synonyms"] = merged[merged.notna()].apply(lambda d: d["synonyms"])
cell_line_lookup.loc[resolved_mask, "_resolution_step"] = "stripped_name_match"

# ---- Step 2d: synonym fallback (exploded, first match wins) ------------
unmatched = cell_line_lookup["cvcl_accession"].isna()
syn = ces[["cellosaurus_accession", "synonyms"]].copy()
syn["synonyms_exploded"] = syn["synonyms"].astype(str).str.split("|")
syn = syn.explode("synonyms_exploded")
syn["synonyms_exploded"] = syn["synonyms_exploded"].str.strip()
syn = syn.dropna(subset=["synonyms_exploded"]).drop_duplicates("synonyms_exploded")
syn_map = syn.set_index("synonyms_exploded")[["cellosaurus_accession", "synonyms"]].to_dict("index")

merged = cell_line_lookup.loc[unmatched, "cell_line_name"].map(syn_map)
resolved_mask = unmatched & merged.notna()
cell_line_lookup.loc[resolved_mask, "cvcl_accession"] = merged[merged.notna()].apply(lambda d: d["cellosaurus_accession"])
cell_line_lookup.loc[resolved_mask, "synonyms"] = merged[merged.notna()].apply(lambda d: d["synonyms"])
cell_line_lookup.loc[resolved_mask, "_resolution_step"] = "synonym_fallback"

# ---- Step 2e: Sanger xref match -----------------------------------------
unmatched = cell_line_lookup["cvcl_accession"].isna()
sanger_lookup = ces.dropna(subset=["sanger_xref"]).drop_duplicates("sanger_xref")
sanger_map = sanger_lookup.set_index("sanger_xref")[["cellosaurus_accession", "synonyms"]].to_dict("index")

merged = df_sample_info_cleaned.set_index("depmap_id")["sanger_model_id"].reindex(
    cell_line_lookup.loc[unmatched, "model_id"]
).map(sanger_map)
merged.index = cell_line_lookup.loc[unmatched].index
resolved_mask = unmatched & merged.notna()
cell_line_lookup.loc[resolved_mask, "cvcl_accession"] = merged[merged.notna()].apply(lambda d: d["cellosaurus_accession"])
cell_line_lookup.loc[resolved_mask, "synonyms"] = merged[merged.notna()].apply(lambda d: d["synonyms"])
cell_line_lookup.loc[resolved_mask, "_resolution_step"] = "sanger_xref"

cell_line_lookup.drop(columns=["_rrid_cvcl"], inplace=True)

# ---- Unmapped log — cellosaurus side (§5 spec) ---------------------------
unmapped_cellosaurus = cell_line_lookup.loc[
    cell_line_lookup["cvcl_accession"].isna(), ["model_id"]
].copy()
unmapped_cellosaurus["dataset"] = "cellosaurus"
unmapped_cellosaurus["resolution_step_reached"] = "cascade_exhausted"
unmapped_cellosaurus["reason"] = "no match at RRID / secondary accession / name / synonym / sanger_xref"
unmapped_cellosaurus = unmapped_cellosaurus.rename(columns={"model_id": "original_value"})[
    ["original_value", "dataset", "resolution_step_reached", "reason"]
]

Index(['cellosaurus_cell_line_name', 'cellosaurus_accession',
       'secondary accession number(s)', 'synonyms', 'cross-references',
       'references identifiers', 'web pages', 'comments', 'str profile data',
       'diseases', 'species of origin', 'hierarchy',
       'originate from same individual', 'sex of cell',
       'age of donor at sampling', 'category', 'date (entry history)',
       'refs_parsed', 'depmap_xref', 'cosmic_xref', 'sanger_xref'],
      dtype='str')


In [146]:
# depmap_profiles_cleaned_pivoted.modelid is ALREADY ACH-format direct key match,
# no cascade required. NOT part of cell_line_lookup's fixed schema kept as a
# separate enrichment frame so the core lookup table's contract stays untouched.

cell_line_lookup_with_profiles = cell_line_lookup.merge(
    df_depmap_profiles_cleaned_pivoted,
    left_on="model_id",
    right_on="modelid",
    how="left",
    indicator=True,
)

# Orphaned direction: profiles rows whose modelid has no home in cell_line_lookup.
_profile_orphans = df_depmap_profiles_cleaned_pivoted.loc[
    ~df_depmap_profiles_cleaned_pivoted["modelid"].isin(cell_line_lookup["model_id"])
]
unmapped_depmap_profiles = _profile_orphans[["modelid"]].copy()
unmapped_depmap_profiles["dataset"] = "depmap_profiles"
unmapped_depmap_profiles["resolution_step_reached"] = "direct_join"
unmapped_depmap_profiles["reason"] = "modelid not present in cell_line_lookup.model_id"
unmapped_depmap_profiles = unmapped_depmap_profiles.rename(columns={"modelid": "original_value"})[
    ["original_value", "dataset", "resolution_step_reached", "reason"]
]

cell_line_lookup_with_profiles.drop(columns=["modelid", "_merge"], inplace=True)

print("cell_line_lookup:", len(cell_line_lookup), "rows")
print(cell_line_lookup["_resolution_step"].value_counts(dropna=False))
print()
print("Unmapped cellosaurus rows:", len(unmapped_cellosaurus))
print("Unmapped depmap_profiles rows:", len(unmapped_depmap_profiles))

cell_line_lookup: 1840 rows
_resolution_step
name_match             1391
stripped_name_match     299
sanger_xref              76
<NA>                     68
synonym_fallback          6
Name: count, dtype: int64

Unmapped cellosaurus rows: 68
Unmapped depmap_profiles rows: 73


In [147]:
# cell line lookup overview 
cell_line_lookup_with_profiles.head()

,model_id,cell_line_name,stripped_cell_line_name,rrid,cvcl_accession,synonyms,_resolution_step,has_rna,has_wes,has_wgs,modelcondition_rna,modelcondition_wes,modelcondition_wgs,weskit_wes
0,ach-000016,slr 21,slr21,cvcl_v607,cvcl_v607,slr 21,stripped_name_match,True,False,False,mc-000016-0cgg,NaN,NaN,NaN
1,ach-000032,mhh-call-3,mhhcall3,cvcl_0089,cvcl_0089,mhh-call 3; mhh call 3; mhh-call3; mhh-call3; ...,name_match,True,True,False,mc-000032-b8um,mc-000032-b8um,NaN,agilent
2,ach-000033,nci-h1819,ncih1819,cvcl_1497,cvcl_1497,h1819; h-1819; ncih1819,name_match,True,False,False,mc-000033-96oj,NaN,NaN,NaN
3,ach-000043,hs 895.t,hs895t,cvcl_0993,cvcl_0993,hs-895-t; hs 895 t; hs895.t; hs895t; hs895t,name_match,True,True,False,mc-000043-0ycn,mc-000043-0ycn,NaN,agilent
4,ach-000049,hek te,hekte,cvcl_ws59,cvcl_ws59,hek-te; hekte,name_match,True,False,False,mc-000049-k0ty,NaN,NaN,NaN


In [148]:
# unmapped log details of cellosaurus 
unmapped_cellosaurus.head()

,original_value,dataset,resolution_step_reached,reason
8,ach-000071,cellosaurus,cascade_exhausted,no match at RRID / secondary accession / name ...
21,ach-000154,cellosaurus,cascade_exhausted,no match at RRID / secondary accession / name ...
23,ach-000165,cellosaurus,cascade_exhausted,no match at RRID / secondary accession / name ...
26,ach-000172,cellosaurus,cascade_exhausted,no match at RRID / secondary accession / name ...
28,ach-000175,cellosaurus,cascade_exhausted,no match at RRID / secondary accession / name ...


In [149]:
# unmapped details of the depmap profile 
unmapped_depmap_profiles.head()

datatype,original_value,dataset,resolution_step_reached,reason
1195,ach-001437,depmap_profiles,direct_join,modelid not present in cell_line_lookup.model_id
1196,ach-001438,depmap_profiles,direct_join,modelid not present in cell_line_lookup.model_id
1210,ach-001481,depmap_profiles,direct_join,modelid not present in cell_line_lookup.model_id
1221,ach-001508,depmap_profiles,direct_join,modelid not present in cell_line_lookup.model_id
1244,ach-001537,depmap_profiles,direct_join,modelid not present in cell_line_lookup.model_id


### INSIGHTS FROM THE HARMONIZATION 
- Comprehensive Tracking: The dual unmapped logging approach provides visibility into both directions of the mapping problem - what couldn't be mapped to
  reference standards AND what reference data couldn't be integrated.
-  Data Quality Indicators: The 68 unmapped cell lines (3.7%) may represent:
    - Novel or poorly characterized cell lines
    - Data quality issues in source identifiers
    - Limitations in current cross-reference databases
    - Cell lines requiring manual curation
- Process Transparency: By recording the specific reason and resolution step reached for each unmapped record, the process enables targeted follow-up and continuous improvement of the matching algorithms.
- Integration Completeness: The high success rate (96.3%) demonstrates the effectiveness of the multi-identifier cascading approach, with direct name matching being the most productive single strategy.

In [150]:
# log the directory 
logs_dir = "logs"  
os.makedirs(logs_dir, exist_ok=True)

# Save unmapped cellosaurus log
cellosaurus_log_path = os.path.join(logs_dir, 'trackA_cellosaurus_unmapped_log.csv')
print(cellosaurus_log_path)
unmapped_cellosaurus.to_csv(cellosaurus_log_path, index=False)
  
# Save unmapped depmap profiles log
depmap_log_path = os.path.join(logs_dir, 'trackA_depmap_profiles_unmapped_log.csv')
print(depmap_log_path)
unmapped_depmap_profiles.to_csv(depmap_log_path, index=False)

print("saved the logged files !!!")

logs\trackA_cellosaurus_unmapped_log.csv
logs\trackA_depmap_profiles_unmapped_log.csv
saved the logged files !!!


In [151]:
# Store the cell_line_lookup_with_profiles as the paraquet under the data 
harmonization_path = "data/parquet/harmonized"
os.makedirs(harmonization_path, exist_ok=True)

# save the cell_line_lookup_with_profiles in trackA_harmonization.parquet 
trackA_harmonization_path = os.path.join(harmonization_path, 'trackA_harmonization.parquet')
print(trackA_harmonization_path)
cell_line_lookup_with_profiles.to_parquet(trackA_harmonization_path, index=False)

data/parquet/harmonized\trackA_harmonization.parquet
